### DISCUSSION TITLE
Week 2: June 24-28

### DISCUSSION THREAD

**Discussion by eeholmes** [SAFS-Varanasi-Internship/Summer-2024#12] [General]
@ShridharS19 @yifeihang notes here

**Comment by eeholmes** [comment]
Monday:

* Eli: Got GPU for JupyterHub but crashed the hub.
* Shridhar: Got PCA model going but dhub is running out of RAM.
*

**Comment by yifeihang** [comment]
![image](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/c50ee8ff-f843-4b53-8540-6aaea14fca22)

**Comment by yifeihang** [comment]
log(CHL)
![image](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/36a3e7b5-52fa-4a14-8fdb-d4059a11fda7)

**Comment by ShridharS19** [comment]
06/24/2024 - Updates
- Running Random Forest Predictor on 1-month slices of 8-year data time series. If the zarr file is shifted to the fish hub(more RAM) and GPU access is available then the predictor can be trained on larger data.
- 
![renljngeljrn](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/108433641/c8266b2c-7b2f-4abd-be77-8e60bc343418)
- Trying to do PCA on chlorophyll data to understand dependence on Principal components and get good reconstruction photos.
- Found a reference here: https://www.sciencedirect.com/science/article/pii/S0025326X20300205?via=ihub

**Comment by yifeihang** [comment]
06/24/24 (Mon.)

Progress:
- Got GPU running locally and ran the ConvLSTM tutorial
- Worked on the CUDA installation tutorial
- Tried to create a multi-variable ConvLSTM model for CHL prediction. Current progress: data preprocessing

Question about the ConvLSTM_Tutorial: @NaNa7Miiii 
- For function "preprocess_data", why is chunking necessary given that the data was already in chunks in the zarr dataset? Also, I found inconsistent chunk sizes across variables in the zarr dataset, will the rechunking process you did in your tutorial help with that?
- When running your tutorial, I found that the validation loss is always lower than the training loss at the end of the training process. I tried different data splitting orders and removing dropout layers, but neither made a huge difference. I thought train loss is usually an underestimate of the true loss, so why would this model be the reverse?

Goal for tomorrow:
- Get the multi-variable ConvLSTM model working
- Continue working on the CUDA installation tutorial

**Comment by eeholmes** [comment]
CHL

https://coastwatch.noaa.gov/erddap/griddap/noaacwNPPN20S3ASCIDINEOF2kmDaily.graph?chlor_a%5B(2020-08-03T12:00:00Z)%5D%5B(0.0)%5D%5B(26.69792):(-6.947917)%5D%5B(51.67709):(85.34376)%5D&.draw=surface&.vars=longitude%7Clatitude%7Cchlor_a&.colorBar=%7C%7C%7C%7C%7C&.bgColor=0xffccccff

<img width="395" alt="image" src="https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/2545978/3572dcf9-8dde-4823-bfb9-8b323ac097c8">


<img width="349" alt="image" src="https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/2545978/6a8fd548-7004-4a72-890e-4f658da12345">

**Comment by ShridharS19** [comment]
06/25/2024
-  Ran into errors while pre-processing for PCA
- Read more closely into the research paper linked yesterday
To do:
- Move stuff over to fish hub
- Fix pre-processing errors

**Comment by yifeihang** [comment]
06/25/24 (Tues.)

Progress:
- Ran a basic CNN model using two variables (SST and salinity)

![image](https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/141bdcbb-6f7d-48d2-b149-b5a235dbb3b8) <img width="372" alt="image" src="https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/c1628255-259d-4df6-add0-e3639f3a7cbc">

- Ran a ConvLSTM model using three variables with a window size of 5 (SST, salinity, and CHL concentration of the past five days), outperforms the CNN model (but it might be copying from inputs)

<img width="440" alt="image" src="https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/0dfa67b0-27a2-48ad-b9bf-d09ace1f2ae2"> <img width="365" alt="image" src="https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/166258283/ce34f588-7f2b-4a2c-a472-230dd0b44323">



Goal for tomorrow:
- Excluding CHL from input for future models
- Do research on how masking in tensorflow works, and see how it helps with NaN values
- Try how topo affects the performance

If I have more time:
- New variables: wind_speed, curr_speed, mlotst
- Research on how GPU can be detected in JupyterHub (maybe how to download CUDA?
- CUDA tutorial

**Comment by eeholmes** [comment]
Wednesday.

**Eli**

@ShridharS19 I got the GPU working I think.  Currently I just have the Pangeo Pytorch image available. Let me know all the packages that you will need and I will try to find a better image.

@yifeihang Can you point me to a jupyter notebook that I can use to test your code? Do you have one in your repo?

<img width="859" alt="image" src="https://github.com/SAFS-Varanasi-Internship/Summer-2024/assets/2545978/fedb87fe-8371-4947-9849-d70fa6139f3e">

**Reply by yifeihang** [reply]
I pushed what I had yesterday to my repo just now, but I don't see the zarr file or the mask file in the shared folder of fish.opensci (so the file paths are still my local paths). Do you want me to upload them?

**Reply by ShridharS19** [reply]
Sorry I just saw this, I tried uploading last night before sleeping but my laptop crashed, I'm trying again.

**Reply by ShridharS19** [reply]
The zip file should be up

**Reply by eeholmes** [reply]
**Problems with the config file**

This is working:
```
        profile_options:
          image:
            display_name: Image
            choices:
              pytorch:
                display_name: Pangeo PyTorch ML Notebook
                default: true
                slug: "pytorch"
                kubespawner_override:
                  image: "quay.io/pangeo/pytorch-notebook:2023.09.19"
```
But if I change the image at all then I get a time out error and the hub will not restart. I want to use this
```
         profile_options:
          image:
            display_name: Image
            choices:
              pytorch:
                display_name: Pangeo PyTorch ML Notebook
                default: true
                slug: "pytorch"
                kubespawner_override:
                  image: "quay.io/pangeo/pytorch-notebook:2024.06.24"
             tensorflow:
                display_name: Pangeo TensorFlow ML Notebook
                default: true
                slug: "tensorflow"
                kubespawner_override:
                  image: "quay.io/pangeo/ml-notebook:2024.06.24"
```

**Comment by yifeihang** [comment]
06/26/24 (Wed.)
Progress - Tried ConvLSTM with: 
1. variables SST and salinity with a 5-day window size: a lot worse than the CNN model from yesterday
2. variables SST, salinity, and topo with a 3-day window: better than model 1
3. SST, salinity, topo, and mlotst with a 3-day window: similar to model 2

All failed to capture any intricate patterns in CHL, and all are worse than the CNN model

Goal for tomorrow:
1. Try adding: ['wind_dir', 'wind_speed'] or ['u_wind', 'v_wind']
2. Work on slides for Friday presentation
3. Maybe try CNN (without LSTM)

**Comment by yifeihang** [comment]
06/27/24 (Thurs.)
Progress:
- Tried 3-day window ConvLSTM with variables:
i. u_wind and v_wind: better than yesterday's models
ii. wind_dir and wind_speed: worse than the first model
iii. model 1 with topo: had a nice plot for losses, but the prediction seems more off (has an obvious cyan boundary around the water body)
iiii. model 1 with other variables: bad models, make all-zero predictions
May due to different units/scales across features

&nbsp;&nbsp;&nbsp;&nbsp; Also, unable to train a model with 4 variables or more due to memory restriction
- Worked on the slides for tomorrow

Goal for tomorrow:
- Run the uwind+vwind model across the whole year, record the losses and plot them
- Try standardization (data for each feature will have 0 mean and 1 standard deviation)
- Try custom loss function (that excludes the NaN regions when calculating loss)

**Comment by ShridharS19** [comment]
06/27/2024
- Got the zarr file on fishhub
- PCA Working! Experimenting with masking for land
To do:
- Manually remove the land from the data for PCA and add itback in after reconstruction
- Now that zarr file is up get the Random forest running on larger timescales on fishhub
